[WIP] Redesign capabilities dictionary

mariaschuld opened the pull request:
This PR implements two goals:

1. Redesign of the capabilities dictionary

* The dictionary is now inherited from the parent class
* The keys have more expressive names, indicating the value type (i.e., ``supports_XYZ`` instead of ``XYZ``)
* All keys, plus a few more that will be handy in future, are defined in the base device class
* The developer documentation is updated 

2. Test capabilities (plus supported operations) of a device in the shared device tests

* The supported operations and observables of a device are now tested
* Most of the capabilities of a device are tested

In the process, a few bugs were uncovered, and the ``default.tensor`` device got the attribute ``analytic`` to allow some more tests to run.

mariaschuld commented on pennylane/_device.py:
```diff
@@ -51,7 +51,22 @@ class Device(abc.ABC):
+        'performs_noisy_computation': False,
+        }
+    """The capabilities dictionary stores the properties of a device. Devices can add their 
+    own custom properties and overwrite existing ones by overwriting the ``capabilities`` class method."""
```
Not sure if one would use docstrings for class attributes?

mariaschuld commented on pennylane/_device.py:
```diff
@@ -51,7 +51,22 @@ class Device(abc.ABC):
+        'supports_exact': False,
+        'supports_sampled': False,
+        'supports_inverse_operations': False,
+        'supports_tensor_observables': False,
```
I decided to go with "supports" since it gives a clear hint at the boolean type of the value.

mariaschuld commented on pennylane/_device.py:
```diff
@@ -51,7 +51,22 @@ class Device(abc.ABC):
+        'provides_jacobian': False,
+        'executes_in_remote': False,
+        'takes_fixed_number_of_wires': False,
+        'performs_noisy_computation': False,
```
Added this last key, because it can be useful for tests. Noise is different to stochasticity, because a circuit that produces a quantum state |01001> will return a deterministic result on a stochastic/sampled simulator, but not on a noisy hardware device or simulator.

mariaschuld commented on pennylane/_qubit_device.py:
```diff
@@ -111,32 +111,14 @@ def __init__(self, wires=1, shots=1000, analytic=True):
-        """
-        capabilities = cls._capabilities
-        capabilities.update(model="qubit", tensor_observables=True)
+        capabilities = super().capabilities().copy()
```
I tried outsourcing this line to the base device class in a function that can be inherited, but it creates a big mess with the ``super()`` function. So I left it here.

mariaschuld commented on pennylane/beta/devices/default_tensor.py:
```diff
@@ -165,6 +164,17 @@ def __init__(self, wires, shots=1000, representation="exact", contraction_method
+            supports_sampled=True,
+            supports_tensor_observables=True,
+        )
+        return capabilities
```
Here the new logic is very apparent: defining  ``_capabilities`` creates a new dictionary, while overriding ``capabilities()`` this way will "inherit" all keys defined by the parent classes.

mariaschuld commented on pennylane/beta/devices/default_tensor_tf.py:
```diff
@@ -174,6 +168,15 @@ def __init__(self, wires, shots=1000, representation="exact", contraction_method
+    def capabilities(cls):
+        capabilities = super().capabilities().copy()
+        capabilities.update(
+            provides_jacobian=True,
```
"tensor_observables" not needed because it is inherited from ``QubitDevice``.

antalszava replied:
Wait, doesn't `DefaultTensor` inherit from `Device`? (hence `tensor_observables` would not be defined)

mariaschuld replied:
Correct, sorry :)

mariaschuld commented on pennylane/devices/__init__.py:
```diff
@@ -32,3 +32,5 @@
 from .default_qubit import DefaultQubit
 from .default_gaussian import DefaultGaussian
+from .default_qubit_tf import DefaultQubitTF
+from .default_qubit_autograd import DefaultQubitAutograd
```
Did we just forget to add these, or was there a reason?

josh146 replied:
No this is intentional! `DefaultQubit` and `DefaultGaussian` have always been here, but not including `DefaultQubitTF` and `DefaultQubitAutograd` is deliberate. Including them here causes an 'import chain' on `import pennylane`, leading to an `import tensorflow` and an `import autograd` statement.

TensorFlow is an optional requirement, so we don't want this to happen (we only want `default.qubit.tf` to be imported when the user explicitly loads it)

josh146 replied:
It might be good to either 

1. leave a comment here explaining that devices with optional requirements are not imported here, or

2. remove all device imports (but this means updating import paths from
  ```python
   from pennylane.devices import DefaultQubit
  ```
  to 
  ```python
  from pennylane.devices.default_qubit import DefaultQubit
  ```

mariaschuld commented on pennylane/_device.py:
```diff
@@ -51,7 +51,22 @@ class Device(abc.ABC):
-    _capabilities = {}  #: dict[str->*]: plugin capabilities
+    _capabilities = {
+        'model': None,
+        'passthru_interface': None,
```
In two minds about this one: we could only define it in the devices that actually have passthru functionality (so that absence of the key signifies that the device does not)...

mariaschuld commented on pennylane/_qubit_device.py:
```diff
@@ -111,32 +111,14 @@ def __init__(self, wires=1, shots=1000, analytic=True):
-            dict[str->*]: results
-        """
-        capabilities = cls._capabilities
-        capabilities.update(model="qubit", tensor_observables=True)
```
Docstring not needed since method inherited from parent.

mariaschuld commented on pennylane/devices/tests/test_gates.py:
```diff
@@ -33,6 +35,58 @@
+    # 'Squeezing': qml.Squeezing(0, 0, wires=[0]),
+    # 'ThermalState': qml.ThermalState(0, wires=[0]),
+    # 'TwoModeSqueezing': qml.TwoModeSqueezing(0, 0, wires=[0, 1]),
+}
```
Note: It is not terribly essential that this list is 100% maintained. If a gate is missing, we simply don't test here that a device claiming to implement it really implements it. 

A misrepresentation like that is likely to be reported by users.

mariaschuld replied:
Also, I added CV gates, since the test suite will soon be expanded to cv.

antalszava replied:
(Minor suggestion): Would it be worth leaving the CV gates out for now? Perhaps they might be separated into another file

mariaschuld commented on pennylane/devices/tests/test_measurements.py:
```diff
@@ -211,10 +264,10 @@ def test_paulix_pauliy(self, device, tol, skip_if):
             qml.CNOT(wires=[0, 1])
             qml.CNOT(wires=[1, 2])
-            return qml.expval(qml.PauliX(wires=0)), qml.expval(qml.PauliX(wires=2))
+            return qml.expval(qml.PauliX(wires=0) @ qml.PauliY(wires=2))
```
Bug uncovered...

josh146 replied:
How come this test was passing before? 😨

mariaschuld replied:
I think there was never a device natively supporting tensor observables that we ran the test suite with.

josh146 replied:
Wait, wouldn't this run on `default.qubit`?

mariaschuld replied:
Interesting question - but default.qubit did not previously have `tensor_observables` set to True!

josh146 replied:
It inherited it from ``QubitDevice``, I believe:

```python
    @classmethod
    def capabilities(cls):
        capabilities = cls._capabilities
        capabilities.update(model="qubit", tensor_observables=True)
        return capabilities
```

So it should have been running?

mariaschuld replied:
I changed the default value for ``tensor_observables`` in ``QubitDevice``, it was ``False`` :)  But the logic is implemented in the device, so I think this is better.

mariaschuld commented on pennylane/devices/tests/test_measurements.py:
```diff
@@ -449,11 +498,11 @@ def circuit():
         mean = np.mean(res)
         expected = -(np.cos(varphi) * np.sin(phi) + np.sin(varphi) * np.cos(theta)) / np.sqrt(2)
-        assert np.allclose(mean, expected, atol=tol(dev.analytic))
+        assert np.allclose(mean, expected, atol=tol(False))
```
These were also bugs uncovered.

mariaschuld commented on pennylane/devices/tests/test_measurements.py:
```diff
@@ -498,10 +547,10 @@ def circuit():
         Z = np.diag([1, -1])
         eigvals = np.linalg.eigvalsh(np.kron(Z, A_))
-        assert np.allclose(sorted(list(set(res))), sorted(eigvals), atol=tol(dev.analytic))
+        assert np.allclose(sorted(np.unique(res)), sorted(eigvals), atol=tol(False))
```
Autograd complains otherwise, because it returns a tensor.

mariaschuld commented on pennylane/devices/tests/test_measurements.py:
```diff
@@ -468,13 +517,13 @@ def test_hermitian(self, device, tol, skip_if):
         varphi = -0.543
 
-        A_ = np.array(
+        A_ = 0.1 * np.array(
```
This (and below) rescales the Hermitian to effectively rescale the result, so we can use ``atol``.

josh146 replied:
Awesome!

mariaschuld commented on pennylane/devices/tests/test_properties.py:
```diff
@@ -52,3 +45,86 @@ def test_no_0_shots(self, device_kwargs):
+            def circuit():
+                return qml.expval(qml.Identity(wires=0))
+
+            assert hasattr(circuit, "jacobian")
```
These capabilities are currently not tested:

* 'executes_in_remote' (awaiting solution to improve the device tests for remote devices)
* 'takes_fixed_number_of_wires' (awaiting solution to improve the device tests for fixed-wire devices)
* 'performs_noisy_computation' (awaiting solution to improve device tests stochastic cases)

mariaschuld commented on pennylane/beta/devices/default_tensor.py:
```diff
@@ -164,6 +164,18 @@ def __init__(self, wires, shots=1000, representation="exact", contraction_method
         self._rep = representation
         self._contraction_method = contraction_method
         self.reset()
+        self.analytic = False
```
Is it ok if I add this attribute? It does not change anything, but allows to run some more device tests. The final decision of how the ``analytic`` attribute will be set/renamed/used is still outstanding, but an overall strategy is in our near-term goals.

josh146 replied:
:+1:

co9olguy replied:
:rofl:

josh146 commented on pennylane/_device.py:
```diff
@@ -51,7 +51,22 @@ class Device(abc.ABC):
+        "takes_fixed_number_of_wires": False,
+        "performs_noisy_computation": False,
+    }
+    """The capabilities dictionary stores the properties of a device. Devices can add their
+    own custom properties and overwrite existing ones by overriding the ``capabilities()`` method."""
```
I like the use of attribute docstrings :)

josh146 commented on pennylane/_device.py:
```diff
@@ -213,9 +228,19 @@ def map_wires(self, wires):
+            def capabilities(cls):
+                capabilities = super().capabilities().copy()
+                capabilities.update(
+                    new_capability=...,
```
Might be good to fill this in with a minimal example, so that it is copy and pastable

josh146 commented on pennylane/_device.py:
```diff
@@ -213,9 +228,19 @@ def map_wires(self, wires):
+        Inheriting classes that change or add capabilities must override this method, for example via
 
-        Measurements, batching etc.
+        .. code-block::
```
```suggestion
        .. code-block:: python
```

josh146 commented on pennylane/_device.py:
```diff
@@ -413,7 +438,9 @@ def supports_operation(self, operation):
             if operation.endswith(Operation.string_for_inverse):
```
I will be glad to get rid of this once Device no longer sends operation strings!

josh146 commented on pennylane/_device.py:
```diff
@@ -413,7 +438,9 @@ def supports_operation(self, operation):
                 return operation[
                     : -len(Operation.string_for_inverse)
-                ] in self.operations and self.capabilities().get("inverse_operations", False)
+                ] in self.operations and self.capabilities().get(
+                    "supports_inverse_operations", False
+                )
```
Woah, black left this looking really weird

josh146 replied:
Slight tangent, but I've always thought the way we do inverse operations is a bit weird. Really, there are only two gates that need the plugin to implement inverse; S and T. Every single other gate is either its own inverse (PauliZ, Hadamard), or the inverse is a function of the arguments (RX(-x), QubitUnitary(U.conj().T)).

I've always wondered if we could treat inverses like decompositions:

* PL-Core asks the device to implement an inverse of a gate.

* If the device doesn't support the inverse natively, PL fallsback to the above parameter transform, or if unknown, simply raises an error.

josh146 replied:
Having said that, I don't think it's priority to fix, but would be a nice to have.

mariaschuld replied:
Nice idea indeed!

josh146 commented on pennylane/devices/default_gaussian.py:
```diff
@@ -656,8 +656,6 @@ class DefaultGaussian(Device):
     version = "0.12.0"
     author = "Xanadu Inc."
 
-    _capabilities = {"model": "cv"}
```
Just double checking, but as far as I can tell, devices can (if they want to) simply keep defining the `_capabilities` attribute, and things will continue to work as before? That is, this change in `default.gaussian` isn't _required_ to avoid breakages?

mariaschuld replied:
Good point...in theory yes, the external devices simply inherit/overwrite the base classes' ``_capabilities``.

However, I ran the device tests on some of them, and there are a few issues:
* The rename from `inverse_operations` to `supports_inverse_operations` means that even though a device says it supports inverse operations, it registers it under the wrong keyword and the core will not know it. We could fix this by revoking the change for now.
* `QubitDevice` is set by default to support sampled and exact mode. This needs to be adapted for every device inheriting from it. However, these capabilities are only used in the device tests, so core should still work.

mariaschuld replied:
Hmm, in the interest of not breaking anything I will change all core calls to these three keys so that they also look for the old versions.

antalszava replied:
Still a bit confused about this change here. Should it rewrite `_capabilities` that is being inherited with `"model": "cv"` in it?

josh146 commented on pennylane/devices/tests/test_gates.py:
```diff
@@ -17,12 +17,15 @@
 """
 # pylint: disable=no-self-use
 # pylint: disable=too-many-arguments
+# pylint: disable=pointless-statement
```
😆

josh146 commented on pennylane/devices/tests/conftest.py:
```diff
@@ -104,7 +104,7 @@ def _device(wires):
 
         capabilities = dev.capabilities()
-        if "model" not in capabilities or not capabilities["model"] == "qubit":
+        if capabilities.get("model", None) != "qubit":
```
Just thinking out-loud here: if we had capability validation in PL-core, we could avoid devices with invalid capabilities being loaded in the first place. E.g., `qml.device("my.device")` would raise an error if `MyDevice.capabilities()["model"] not in ("qubit", "cv")`. This might be a good thing, since it avoids PL loading devices that could lead to unexpected/broken behaviour.

There are two places we could put a validation check:

* `qml.device`

* The QNode itself

mariaschuld replied:
I put it on to my list!

josh146 commented on pennylane/devices/tests/test_gates.py:
```diff
@@ -150,6 +205,45 @@
 
 
+class TestSupportedGates:
+    """Test that the device can implement all gates that it supports."""
```
```suggestion
    """Test that the device can implement all gates that it claims to support."""
```

josh146 commented on pennylane/devices/tests/test_gates.py:
```diff
@@ -150,6 +205,45 @@
+                ops[operation]
+                return qml.expval(qml.Identity(wires=0))
+
+            assert isinstance(circuit(), (float, tensor))
```
Does
```python
assert isinstance(circuit(), (float, np.ndarray))
```
not work? It should, since `qml.numpy.tensor` is a subclass of `ndarray` (and it will avoid the import above)

mariaschuld replied:
Oh sure!

josh146 reviewed (approved):
💯

mariaschuld commented on pennylane/_device.py:
```diff
@@ -213,9 +227,19 @@ def map_wires(self, wires):
+            def capabilities(cls):
+                capabilities = super().capabilities().copy()
+                capabilities.update(
+                    new_capability=...,
```
```suggestion
                    supports_inverse_operations=False,
                    supports_a_new_capability=True,
```

mariaschuld commented on pennylane/beta/devices/default_tensor.py:
```diff
@@ -215,20 +227,20 @@ def _add_node(self, A, wires, name="UnnamedNode", key="state"):
-           sequences of length :math:`k`.
+         Input states can be factorized or entangled. If a state can be factorized
+         into :math:`k` subsystems, then ``tensors``, ``wires``, and ``names`` should be
+         sequences of length :math:`k`.
```
These changes are from updating "black" to v0.20.

mariaschuld commented on pennylane/_device.py:
```diff
@@ -411,9 +436,11 @@ def supports_operation(self, operation):
+                # TODO: update when all capabilities keys changed to "supports_inverse_operations"
+                supports_inv = (self.capabilities().get("supports_inverse_operations", False) or
+                                self.capabilities().get("inverse_operations", False))
+                return in_ops and supports_inv
```
I know the change makes it a bit lengthy, but I wanted to use a structure where we just have to delete the ``or self.capabilities().get("inverse_operations", False)`` part once the keys are unified across external devices.

mariaschuld commented:
@antalszava would you mind reviewing this one, so we can proceed and merge? I tested as much as I could, and our main devices should still pass the device tests. We therefore have time to implement the changes in the devices. In the meantime, the worst that can happen is that some of their less important capabilities are misreported, but without any consequences in core.

antalszava commented on doc/development/plugins.rst:
```diff
@@ -101,15 +101,35 @@ as well as potential further capabilities, by providing the following class attr
+  *  ``'supports_tensor_observables'`` (*bool*): ``True`` if the device supports observables composed from tensor
+     products such as ``PauliZ(wires=0) @ PauliZ(wires=1)``.
+
+  *  ``'executes_in_remote'`` (*bool*): ``True`` if the device runs computations on remote servers or devices.
```
For my own understanding: what would be a use case for `executes_in_remote`?

mariaschuld replied:
For example when devices are advertised on our homepage, and their capabilities automatically searched through...

antalszava replied:
Got it. Not quite sure if it's best to add an entry in the code itself for documentation purposes (just because it gives the impression that specific logic would follow), although I agree that this seems to be a brief enough solution.

antalszava commented on doc/development/plugins.rst:
```diff
@@ -101,15 +101,35 @@ as well as potential further capabilities, by providing the following class attr
+
+  *  ``'takes_fixed_number_of_wires'`` (*bool*): ``True`` if the device's number of wires are fixed.
+
+  *  ``'runs_noisy_computations'`` (*bool*): ``True`` if the device's results are stochastic due to
+     physical or simulated noise.
```
> if the device's results are stochastic due to physical

Would that mean that hardware devices would have this attribute be set to `True`?

mariaschuld replied:
Yes, and some simulators...

antalszava commented on doc/development/plugins.rst:
```diff
@@ -101,15 +101,35 @@ as well as potential further capabilities, by providing the following class attr
+     physical or simulated noise.
+
+  Capabilities are queried by PennyLane core to make decisions on how to best run computations,
+  and by external apps build on top of the device ecosystem.
```
(let me know if missed something)
```suggestion
  and by external apps built on top of the device ecosystem.
```

antalszava commented on pennylane/_qubit_device.py:
```diff
@@ -113,32 +113,14 @@ def __init__(self, wires=1, shots=1000, analytic=True):
+        capabilities = super().capabilities().copy()
+        capabilities.update(
+            model="qubit",
+            supports_exact=True,
```
Would the idea be that for hardware devices this is explicitly set to `False`? (`IBMQDevice`, `QPUDevice`)

mariaschuld replied:
You are right, maybe better to have default False?

antalszava commented on pennylane/_qubit_device.py:
```diff
@@ -113,32 +113,14 @@ def __init__(self, wires=1, shots=1000, analytic=True):
+        capabilities.update(
+            model="qubit",
+            supports_exact=True,
+            supports_sampled=True,
```
Think the only exception to this would be using the `unitary_simulator` backend from the Qiskit plugin

mariaschuld replied:
Yes, that needs to be updated. But since nothing in PL uses those capabilities yet, we have time to do that!

antalszava commented on pennylane/devices/__init__.py:
```diff
@@ -30,5 +30,8 @@
     autograd_ops
     tests
 """
+# DefaultQubitTF and DefaultQubitAutograd not imported here since this
+# would lead to an automatic import of tensorflow and autograd, which are
+# not PennyLane core dependencies
```
:partying_face:

antalszava commented on pennylane/devices/default_qubit.py:
```diff
@@ -360,6 +359,15 @@ def _get_unitary_matrix(self, unitary):  # pylint: disable=no-self-use
+    def capabilities(cls):
+        capabilities = super().capabilities().copy()
+        capabilities.update(
+            supports_reversible_diff=True,
```
How come this would not be defined among the capabilities in `Device`?

mariaschuld replied:
It is set to False in ``Device``...

antalszava replied:
Oh not sure how I missed that, my bad!

antalszava commented on pennylane/devices/tests/test_gates.py:
```diff
@@ -150,6 +204,49 @@
+    @pytest.mark.parametrize("operation", all_ops)
+    def test_supported_gates_can_be_implemented(self, device_kwargs, operation):
+        """Test that the device can implement all its supported gates."""
+        device_kwargs["wires"] = 3  # maximum size of current gates
```
(Minor suggestion): would it be worth factoring out `3` higher up in this file? Seems to be a constant that appears more than once and could be changed easily later if defined at a single place

mariaschuld replied:
Yes, and we wanted to keep test data as close to tests as possible, so people can understand what is tested...

antalszava commented on pennylane/devices/tests/test_gates.py:
```diff
@@ -150,6 +204,49 @@
+                ops[operation]
+                return qml.expval(qml.Identity(wires=0))
+
+            assert isinstance(circuit(), (float, np.ndarray))
```
Just curious, what would be a specific case for this assertion to fail?

mariaschuld replied:
Good question, the assert statement is more of a dummy here. It's one of those tests where I just want to know "does the execution throw an error". Is there a generic way to do this with pytest? One could also not have an assert statement at all, as in some other test, but I thought this is bad practice...

antalszava reviewed (approved):
Looks good to me, great on @mariaschuld! :100: Had some questions, mostly for my own understanding

mariaschuld commented:
After double approval I implemented quite a few more changes: 

* The base device class now *only* defines the ``model:None`` capability, everything else is up to the device  
* Added keys ``returns_probs`` and ``returns_state``, and use the former to exclude devices from tests
* Improved the capabilities tests to complain if a capability is misreported (before it only complained if it was illegally claimed to be True)

@trbromley , since the second change is partly requested by you, would you mind having a look at the overall PR too?

trbromley commented on pennylane/devices/tests/test_properties.py:
```diff
@@ -52,3 +87,144 @@ def test_no_0_shots(self, device_kwargs):
+        if cap["returns_state"]:
+            dev.state  # pylint: disable = pointless-statement
+        else:
+            with pytest.raises(AttributeError):
+                dev.state  # pylint: disable = pointless-statement
```
Some devices (e.g. `qiskit.aer`) have the `dev.state` attribute available, but it is equal to `None`. I would say that for such devices, `cap["returns_state"] = False` should be the case. However, I think this test would fail for such a device since it wouldn't raise an `AttributeError` :thinking:

If I understand the layout of the test suite correctly, this test would be applied to `qiskit.aer` in https://github.com/PennyLaneAI/plugin-test-matrix (I guess it would skip for now, until we define the capabilities for plugins).

Maybe we should have
```python
if cap["returns_state"]:
    assert dev.state is not None
else:
    try:
        state = dev.state
    except AttributeError:
        state = None

    assert state is None
```

trbromley commented on tests/beta/test_default_tensor.py:
```diff
@@ -819,6 +819,20 @@ class TestDefaultTensorIntegration:
     """Integration tests for default.tensor. This test ensures it integrates
     properly with the PennyLane interface, in particular QNode."""
 
+    @pytest.mark.parametrize("capabilities", [{"model": "qubit",
+                                               "supports_finite_shots": False,
+                                               "supports_tensor_observables": True,
+                                               "returns_probs": False,
+                                               "returns_state": False,
+                                               "supports_analytic_computation": True,
+                                               }])
```
Is it best practice to pass through `parametrize` for just one object?

mariaschuld replied:
Lol, I didn't even notice :) Yes, that makes no sense

trbromley commented on pennylane/devices/default_gaussian.py:
```diff
@@ -695,6 +693,19 @@ def __init__(self, wires, *, shots=1000, hbar=2, analytic=True):
+            supports_analytic_computation=True,
+            supports_finite_shots=True,
+            returns_probs=False,
+            returns_state=False,
```
Perhaps we could set `returns_state=False` in the base `Device`, and then only turn it on when needed.

mariaschuld replied:
The reason why I reduced the default values in the base device is that not having the value is better than misreporting. Also, it would set this value for all devices, and it would be better to go through them one by one...

trbromley reviewed (commented):
Thanks @mariaschuld! Just left minor comments. One thing I was wondering is whether we should set `return_state=False` in the base `Device`?

antalszava commented on pennylane/devices/tests/test_gates.py:
```diff
@@ -488,6 +500,7 @@ def test_two_qubit_no_parameters(self, device, init_state, op, mat, tol, skip_if
         n_wires = 2
         dev = device(n_wires)
         skip_if(dev, {"supports_inverse_operations": False})
+        skip_if(dev, {"returns_probs": False})
```
seems to have complaints for `theta` in this file too :thinking:

antalszava reviewed (approved):
:100: (no particular suggestions, good to go from my side once Tom's major suggestions are also addressed)

mariaschuld commented:
About https://github.com/PennyLaneAI/pennylane/pull/781#discussion_r487022558:

I test it by checking if one can use the device in a reversible qnode and can compute a gradient...but maybe there is a better option. Luckily, the device test are due for an upgrade.
